pkg/server/plugin/keymanager/awskms/fetcher.go

```diff
@@ -0,0 +1,137 @@
+package awskms
+
+import (
+	"context"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+	"github.com/hashicorp/go-hclog"
+	keymanagerv0 "github.com/spiffe/spire/proto/spire/plugin/server/keymanager/v0"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type keyFetcher struct {
+	log         hclog.Logger
+	kmsClient   kmsClient
+	serverID    string
+	trustDomain string
+}
+
+func (kf *keyFetcher) fetchKeyEntries(ctx context.Context) ([]*keyEntry, error) {
+	var keyEntries []*keyEntry
+	var keyEntriesMutex sync.Mutex
+	paginator := kms.NewListAliasesPaginator(kf.kmsClient, &kms.ListAliasesInput{Limit: aws.Int32(100)})
+	g, ctx := errgroup.WithContext(ctx)
+
+	for {
+		aliasesResp, err := paginator.NextPage(ctx)
+		switch {
+		case err != nil:
+			return nil, status.Errorf(codes.Internal, "failed to fetch aliases: %v", err)
+		case aliasesResp == nil:
+			return nil, status.Errorf(codes.Internal, "failed to fetch aliases: nil response")
+		}
+
+		kf.log.Debug("Found aliases", "num_aliases", len(aliasesResp.Aliases))
+
+		for _, alias := range aliasesResp.Aliases {
+			switch {
```

Comment:
How likely are the errors in this `switch` to happen? Are they just exhaustive `nil` checks or are there any realistic scenarios where they might be encountered?

Any of these errors would cause SPIRE server to fail to startup. An operator of SPIRE would presumably need to take some action to resolve these errors to get the server to come up cleanly, but the error messages we build here don't really provide a clear direction to the operator for what they need to do.

If these error conditions might happen due to some unpredicted operator behavior or a bug in SPIRE, it seems valuable to give the user clear direction on how to get the server up. But if these errors would only happen when KMS is down or not behaving according to the API contract, then I am less concerned.

Reply:
These fields are all optional from what I can tell. I don't think we're handling this correctly. These could be aliases managed by someone other than SPIRE. We shouldn't be failing the list operations because these fields are missing.

Reply:
I do think the errors should describe what's going on better , but I don't think these fields are optional. These fields should never be nil, both alias name and target key id are required when creating an alias. The problem is that I have seen keys being listed with a nil target key id. The first time I listed aliases in regions they came without target keys. On a second listing and from then on they came normally.

Reply:
I think the errors should explain that something bad is going on with the KMS API. The reasoning to fail the init instead of continuing with the next key is that due to failures in the KMS API, we could end up building an inaccurate local representation of the keys that belong to the server.

Reply:
To be clear, the aliases that came nil where aliases create by AWS, and only the first time that region was listed. But that was enough to make want to write defensive code.

Reply:
for reference https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/kms#CreateAliasInput

Reply:
Alternatively we could check for `alias.AliasName` not being nil. Then 
```
spireKeyID, ok := kf.spireKeyIDFromAlias(*alias.AliasName)
// ignore aliases/keys not belonging to this server
if !ok {
continue
}
```

then all the other checks.